### ISSUE TITLE
Electron 2.0 issues with rendering Tray Icon on Debian Stretch

### ISSUE THREAD

**Issue by cifratila** [electron/electron#12791]
* Electron version: 2.0.0-beta.7
* Operating system: Debian Linux 9 (stretch)

* Last known working Electron version: 1.8.x

### Expected behavior

Tray icon should be rendered as expected

### Actual behavior

Check out this: https://answers.microsoft.com/en-us/skype/forum/skype_linux-skype_startms-skype_installms/system-tray-icon-in-xfce/d3f162bf-0bbf-481b-90a1-f43cae9a86cc?messageId=07ac5b21-a4cd-43b6-a8b9-cc79fc3c6c54

### How to reproduce

Install Skype for Linux 8.20.76.1 (insiders).
Haven't tried it in a simple Electron app, but should be easily reproducible if you have a tray icon set.

**Comment by ckerr** [comment]
Screenshot from answers.microsoft.com thread:

![tray-icon](https://user-images.githubusercontent.com/70381/39529844-427b5f90-4ded-11e8-8be7-3c0dd73f9d23.png)

**Comment by hovancik** [comment]
can confirm with my app as well ("electron": "^2.0.0")


![screenshot_2018-05-06_12-31-51](https://user-images.githubusercontent.com/4745859/39672394-bf9558e8-5129-11e8-8560-55210cced035.png)

**Comment by blaatunga** [comment]
Currently reported from:

- Xubuntu 17.10, XFCE
- Debian 9, XFCE
- Fedora 27, MATE
- Fedora, LXDE
- Fedora 27, Cinnamon
- Budgie Desktop

**Comment by akontsevich** [comment]
Also have bugs with Electron app icons in openSUSE - #13000. Works fine in 1.8.x and bad in 2.0.1:
- Riot has purple square
- Skype icon become small

**Comment by Staudey** [comment]
> So note that if you set env property XDG_CURRENT_DESKTOP: Unity, then it seems to work. Reportedly App Indicators are used in that case and it does not fall back to Notication Area applets nor disappear completely.

This creates a very big, blurry tray icon on Budgie though. Not sure which glitch is better. ^^

**Comment by psposito** [comment]
As repeatedly has been reported but may not mentioned, it happens generally on any XFCE, not related directly with any Linux Distribution..... 
Add Arch Linux / Manjaro to the above list too....

**Comment by pendula95** [comment]
Just to add to the list: Fedora 28 XFCE

**Comment by jancborchardt** [comment]
Can confirm on Ubuntu 18.04.1 LTS with GNOME 3.28.2. (Basically latest stable)

**Comment by akontsevich** [comment]
Seems libappindicator-gtk3 (libappindicator3) issue: https://github.com/vector-im/riot-web/issues/6767#issuecomment-424535414

**Comment by phhpro** [comment]
Debian 9.5 x64 LXDE -- Issue persists with appindicator completely removed, so fallback to whatever default equally fails; and no change with ayatana either. If this was a libs thing it should affect other icons as well, but it's only with Skype.

**Comment by Heziode** [comment]
@phhpro, 

For me when I remove `appindicator`, on Archlinux with KDE, it's work.

![tray_icon_without_libappindicator](https://user-images.githubusercontent.com/16735076/48963034-b8d08a00-ef8a-11e8-888e-a7b920c82c7f.png)

Seer vladimiry/email-securely-app#75.

**Comment by hovancik** [comment]
If try to uninstall it, it also wanna uninstall my app, so no way, I guess

**Comment by Heziode** [comment]
Depending to your ditsro, you must have a command that only removed this package without its dependencies.

`pacman -Rdd` on Archlinux, for other distro, check the doc ;)

**Comment by patrakov** [comment]
There is no need to remove libappindicator. The same issue affects Skype, and there, it helps to export one variable (even on desktops other than Unity):

export XDG_CURRENT_DESKTOP=Unity

I.e., in the desktop file, you change the Exec= line like this:

Exec=/usr/bin/env XDG_CURRENT_DESKTOP=Unity PULSE_LATENCY_MSEC=30 /usr/bin/skypeforlinux %U

PULSE_LATENCY_MSEC is unrelated to this bug, but helps Skype not to fall back to the worst possible audio codec.

**Comment by akontsevich** [comment]
@patrakov ha, this works! Thanks! I tried `XDG_CURRENT_DESKTOP=Unity` previously without luck. Will change electron apps shortcuts. However this does not fix tray icon left click issue. To fix need to uninstall libappindicator3 (libappindicator-gtk3), replace or patch like some distros do.

**Comment by akontsevich** [comment]
@phhpro hmm, this worked previously for me. Now it works for skype only. Riot icon is very small for some reason:
![default](https://user-images.githubusercontent.com/3843657/49048068-29251880-f1e2-11e8-99fc-6a6acc560ece.png)
Probably will be fixed on Electron upgrade. Will revert `libappindicator`s back for now.

**Comment by nickdante8** [comment]
Hello!

There is a more adequate workaround which does not make the icon tray so blurry. As mentioned by **Rok Mandeljc**, in this [post](https://goo.gl/qqMzDj), it's more a problem of upscaling, which electron has problem to do so, but it seems it does not have problems downscaling it. So the solution would be changing those 16x16 icons to 32x32.

Here's description of the solution:

> Same issue on Fedora 28 with GNOME desktop.
> 
> Exporting _XDG_CURRENT_DESKTOP=Unity_ before starting skype (and installing libappindicator-gtk3) to force the use of appindicator instead of gtk3 tray icon does fix the issue, but disables the window activation via left-click on tray icon (because appindicator does not distinguish between left and right click on the icon).
> 
> The electron framework appears to have issues with upscaling the tray image, but does not seem to have a problem with downscaling the image. So an alternative work-around for the current situation is to replace the small (16px) images with larger versions - and I suspect that even without this issue, in most cases, downscaling a 256px icon to (arbitrary) target tray size will look nicer than upscaling the extra small icons.
> Until this is fixed in electron or worked around in skype by using larger icons by default, you can work around it yourself. To do so, one needs to unpack /usr/share/skypeforlinux/resources/app.asar (using electron asar tool), replace the images in <unpacked-dir>/app/images/tray/linux/, and pack it back. The tray icons are already provided in two versions, 16px and 32px (HiDPI, @2x name suffix). So if your tray size does not exceed 32px, you can simply copy the 32px versions over 16px ones.
> 
> Thank you! This solved the problem!
> 
> [DETAILED SOLUTION]
> 
> I put an already fixed **app.asar** here, on my Google Drive: [app.asar.](https://drive.google.com/file/d/1jWNVnN0gaaOqghhoa_t0WKPPZka4hAMC/view) It is for Skype version 8.33.0.41 (latest for now).
> 
> Just do "**sudo cp app.asar /usr/share/skypeforlinux/resources/app.asar**", restart your Skype and enjoy the large, shiny icon.
> 
> **If you own a different version of Skype**, you have to repack it yourself, as **Rok Mandeljc** advised. The commands to do this are, as follows:
> 
> $ sudo npm install -g asar
> $ asar extract /usr/share/skypeforlinux/resources/app.asar ./skypeapp
> $ cd ./skypeapp/images/tray/linux
> $ ls -1 | grep "@2x" | while read -r pngFile; do cp "./$pngFile" "./${pngFile//$@@2x/}"; done
> $ cd ../../../../
> $ asar pack ./skypeapp ./app.asar
> $ sudo cp ./app.asar /usr/share/skypeforlinux/resources/app.asar
> 
> Done.

**Comment by kode54** [comment]
Creates two distinct tray icons on Deepin Desktop, installed on Arch.

deepin-dock 4.9.2.1-1

![DeepinScreenshot_select-area_20190328205806](https://user-images.githubusercontent.com/796316/55208704-70c1ae00-519c-11e9-85d0-a6ff06284dc9.png)

**Comment by sofianguy** [comment]
@cifratila is this bug still happening on version 4.x or later?

**Comment by akontsevich** [comment]
@sofianguy yes, happening if `libappindicator3` is not installed. If installed then we have issue #13087

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

If this is still reproducible on a supported version, please open a new issue with any other new information that a maintainer should know.

Thank you for taking the time to report this issue and helping to make Electron better! Your help is appreciated.
